src/Microsoft.ML.Dnn/TensorTypeExtensions.cs

```diff
@@ -3,49 +3,477 @@
+                            return;
+                        }
+                        case TF_DataType.TF_INT16:
+                        {
+                            var src = (short*) tensor.buffer;
+                            for (int i = 0; i < size; i++) destination[i] = ((IConvertible) src[i]).ToString(culture);
+                            return;
+                        }
+                        case TF_DataType.TF_UINT16:
+                        {
+                            var src = (ushort*) tensor.buffer;
+                            for (int i = 0; i < size; i++) destination[i] = ((IConvertible) src[i]).ToString(culture);
+                            return;
+                        }
+                        case TF_DataType.TF_INT32:
+                        {
+                            var src = (int*) tensor.buffer;
+                            for (int i = 0; i < size; i++) destination[i] = ((IConvertible) src[i]).ToString(culture);
+                            return;
+                        }
+                        case TF_DataType.TF_UINT32:
+                        {
+                            var src = (uint*) tensor.buffer;
+                            for (int i = 0; i < size; i++) destination[i] = ((IConvertible) src[i]).ToString(culture);
+                            return;
+                        }
+                        case TF_DataType.TF_INT64:
+                        {
+                            var src = (long*) tensor.buffer;
+                            for (int i = 0; i < size; i++) destination[i] = ((IConvertible) src[i]).ToString(culture);
+                            return;
+                        }
+                        case TF_DataType.TF_UINT64:
+                        {
+                            var src = (ulong*) tensor.buffer;
+                            for (int i = 0; i < size; i++) destination[i] = ((IConvertible) src[i]).ToString(culture);
+                            return;
+                        }
+                        case TF_DataType.TF_DOUBLE:
+                        {
+                            var src = (double*) tensor.buffer;
+                            for (int i = 0; i < size; i++) destination[i] = ((IConvertible) src[i]).ToString(culture);
+                            return;
+                        }
+                        case TF_DataType.TF_FLOAT:
+                        {
+                            var src = (float*) tensor.buffer;
+                            for (int i = 0; i < size; i++) destination[i] = ((IConvertible) src[i]).ToString(culture);
+                            return;
+                        }
+#endif
+                        default:
+                            throw new NotSupportedException();
+                    }
+                }
+            }
+        }
+
+        public static void CopyTo<T>(this Tensor tensor, Span<T> destination) where T : unmanaged
```

Comment:
>public static void CopyTo<T>(this Tensor tensor, Span<T> destination) where T : unmanaged [](start = 8, length = 89)

There is a lot of unsafe code in here that does memory copy/manipulation, have you considered using Span<T>.CopyTo(Span) and Span<T>.Slice(int,int) to achieve the same?

Reply:
I don't see a point, I perform all the checks necessary beforehand. The code is safe. Any use of Span will result in unnecessary overhead.

Reply:
@codemzs I can replace:
`var sdst = (bool*) Unsafe.AsPointer(ref destination.GetPinnableReference());`
With:
`fixed (byte* sdst = MemoryMarshal.Cast<T, byte>(destination))`

They do a similar logic just with more overhead.